### PR TITLE
Make `<[T]>::array_*` methods fail to compile on 0 len arrays

### PR DESCRIPTION
**Pull request by AngelicosPhosphoros**
Methods updated:
* `array_windows`
* `array_chunks`
* `array_chunks_mut`
* `as_chunks`
* `as_chunks_mut`
* `as_chunks_unchecked`
* `as_rchunks`
* `as_rchunks_mut`
* `as_chunks_unchecked_mut`

I implemented this using compile time assertions.

Example compilation error:

```
> rustc +stage1 .\improper_array_windows.rs --crate-type=rlib
error[E0080]: evaluation of `core::slice::<impl [u32]>::array_windows::<0>::{constant#0}` failed
    --> J:\rust_lang\library\core\src\slice\mod.rs:1381:13
     |
1381 |             assert!(N != 0, "window size must be non-zero");
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'window size must be non-zero', J:\rust_lang\library\core\src\slice\mod.rs:1381:13
     |
     = note: this error originates in the macro `$crate::panic::panic_2021` which comes from the expansion of the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)

note: the above error was encountered while instantiating `fn core::slice::<impl [u32]>::array_windows::<0>`
 --> .\improper_array_windows.rs:4:14
  |
4 |     for _ in s.array_windows::<0>(){
  |              ^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

I also added doctests and adjusted documentation.

Relates:
https://github.com/rust-lang/rust/issues/74985
https://github.com/rust-lang/rust/issues/75027

### CODE REVIEW

**Comment by AngelicosPhosphoros** [comment]
There are also few other methods, would change them too.

**Comment by lcnr** [comment]
r? rust-lang/libs

**Comment by AngelicosPhosphoros** [comment]
> while it results in a better user experience than runtime panic, this solution has the issue that it only panics during monomorphization, so `check` builds compile successfully, even for `N == 0`.

Ooops, really `rustc +stage1 .\improper_array_windows.rs --crate-type=rlib  --emit=metadata -Z no-codegen` doesn't complain.
However, this would prevent nightly users writing code that we would broke later by adding bound.

**Comment by AngelicosPhosphoros** [comment]
Also relevant: https://github.com/rust-lang/rust/issues/99682

**Comment by AngelicosPhosphoros** [comment]
@joshtriplett

Hi! Do I need to do something with this PR?

**Comment by nbdd0121** [comment]
Might just do `const { assert!(N != 0); }`?

File: library/core/src/slice/mod.rs [hunk 1]
```diff
@@ -110,6 +110,27 @@ enum Direction {
...
+                    "` must be bigger than zero."
+                ))
+            } else {
+            };
+        }
+
+        let _ = <[$t; $array_size] as CompileAssert>::TRIGGER;
+    }};
+}
+
```
Comment (pitaj) [inline library/core/src/slice/mod.rs:133]: As @nbdd0121 suggested, `const { assert!(N != 0); };` is a much simpler way to achieve this. It does still only result in a post-monomorphization error, but it's much cleaner.

Plus, inline const expressions are in the stabilization process right now: #104087

Comment (AngelicosPhosphoros) [reply library/core/src/slice/mod.rs]: Thank you for suggestion, now it looks much nicer.

**Comment by AngelicosPhosphoros** [comment]
@joshtriplett @the8472
Sorry for delay, I was forced to leave my country and had been some time without suitable equipment to work.

I switched to inline const blocks as requested.

**Comment by scottmcm** [comment]
Post-mono errors like this have the problem that they block compilation *even in dead code*, so you can't write (in a const generic) `if N != 0 { foo.as_chunks::<N>() } else { ... }`.

It's thus not obvious to me whether this is actually better than the runtime panic.

Does this have an ACP where the libs-api team has assented to this direction?

**Comment by the8472** [comment]
Ideally this would be a `where N > 0` bound. But the compile time assert is not a proper substitute for that because bound can be infectious while the assert only fires for concrete instantiations.

**Comment by AngelicosPhosphoros** [comment]
@scottmcm Old docs for the methods mentioned that in future any instantiations with `N == 0` would stop compile. My assertions ensure that we wouldn't broke someones code later when we add premonomorphisation check.

Also, we can always remove checks later if we wish because it would not be breaking changes.

**Comment by nbdd0121** [comment]
> Ideally this would be a `where N > 0` bound. But the compile time assert is not a proper substitute for that because bound can be infectious while the assert only fires for concrete instantiations.

Where bound would stop you writing `if N != 0 { foo.as_chunks::<N>() } else { ... }` at all.

**Comment by the8472** [comment]
> > Ideally this would be a `where N > 0` bound. But the compile time assert is not a proper substitute for that because bound can be infectious while the assert only fires for concrete instantiations.
> 
> Where bound would stop you writing `if N != 0 { foo.as_chunks::<N>() } else { ... }` at all.

You can then have a separate implementation for N = 0. Assuming the trait solver can then figure out that for each N there exists exactly one implementation.

**Comment by CAD97** [comment]
The full-fat solution to working with generic instantiation errors is some sort of static if (C++ calls it "constexpr if"), e.g.

```rust
static if N != 0 {
    foo.as_chunks::<N>() // ...
} else {
    // ...
}
```

where each arm of the static if is only instantiated if the controlling condition (which is constant evaluated) says to take that arm.

With arbitrary requirements and conditions, we probably can't prevent instantiation errors (post-mono errors), but we could probably utilize [pattern restricted types](https://github.com/rust-lang/rust/pull/107606) to effect here (i.e. instead of `N: usize`, `N: usize is 1..`). (Though that might have the additional effect of delaying these APIs by perhaps a decade or more...)

**Comment by AngelicosPhosphoros** [comment]
I would like to merge this PR already because it just gather merge conflicts as it stays.

So I like to address all concerns in a single place:
1. Compilation error happens only during monomorphization (e.g. `cargo build` vs `cargo check`). True, but compile error during monomorphization is still better than runtime panic.
2. If we decide later to support case which @scottmcm brought up (`if N != 0 { my_slice.array_xxx::<N>() } else { do something else }`), we can always relax this check later. Removing const from assertions will not break code with `N != 0`. And const assert added now would prevent breaking users code later if we decide that we don't support this usecase.
3. About allowing `N==0` and making infinite loop from it. I am against this proposition for 2 reasons:
    * It is just confusing. Infinite loops are uncommon thing so they should be made only using `loop {}`. Making iteration over chunks infinite loop would be surprising for programmer almost always.
    * It contradicts existing behaviour of `windows(n)` and `chunks(n)` methods so should not be allowed for sake of consistency.

**TLDR**: We cannot broke any valid code by adding this, we can always remove this later and it would prevent breaking more code if we keep this behaviour.

**Comment by joshtriplett** [comment]
I don't feel comfortable unilaterally approving this.

cc @rust-lang/libs-api

**Comment by BurntSushi** [comment]
Let's FCP this perhaps?

@rfcbot fcp merge

**Comment by BurntSushi** [comment]
@rfcbot concern monomorphization errors

> True, but compile error during monomorphization is still better than runtime panic.

I don't agree.

I don't think we should be litigating whether we want to start introducing more monomorphization errors in this PR. It should probably be done as a separate decision, albeit I'm not sure by what process and by what team.

**Comment by dtolnay** [comment]
I agree with https://github.com/rust-lang/rust/pull/99471#issuecomment-1544685018 and I would strongly prefer not to accept this PR. The following is fine and correct code; we do not need to make it fail to compile.

```rust
fn f<const N: usize>() {
    if N == 0 {
        ...
    } else {
        foo.as_chunks::<N>()...
    }
}

fn main() {
    f::<0>();
}
```

**Comment by BurntSushi** [comment]
Given @dtolnay's comment above, the FCP seems unnecessary.

@rfcbot fcp cancel

**Comment by AngelicosPhosphoros** [comment]
> > True, but compile error during monomorphization is still better than runtime panic.
> 
> I don't agree generally.

Maybe allow such compile errors for unstable functions/traits/structs until we implement bounds on constant generics?

**Review by dtolnay — COMMENTED** [review]
Thanks anyway for the PR! This was worth exploring.

**Comment by bend-n** [comment]
what about using ```rust
where [(); N - 1]:
```? im pretty sure this isnt a mono error, which means it works in check builds?
